core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -3421,7 +3421,17 @@ protected boolean checkUserPasswordValid(Object credential) throws UserStoreExce
 
             String regularExpression =
                     realmConfig.getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_JAVA_REG_EX);
-            return regularExpression == null || isFormatCorrect(regularExpression, credentialObj.getChars());
+            if (regularExpression != null) {
+                if (isFormatCorrect(regularExpression, credentialObj.getChars())) {
+                    return true;
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Submitted password does not match with the regex " + regularExpression);
+                    }
+                    return false;
+                }
+            }
+            return regularExpression == null;
```

Comment:
Shall we just say `return true;`. Since the **if** statement will return a result for not null case, at this point **regularExpression** would be null.